docs/cvm/ndarray.rst

```diff
@@ -0,0 +1,96 @@
+NDArray
+=============
+
+This is the document of ``NDArray``, multi-dimension array containing data, a
+core module in CVM Runtime. This part of documentation helps with understanding
+the design and APIs of ``NDArray`` and related classes or methods.
+
+Overview
+--------
+
+``NDArray``, as discussed above, is the multi-dimension array containing data.
+This class provides users with memory management and persistence. The core
+function is implemented with C++ and there are APIs for other languages.
+
+Memory Layout
+-------------
+
+Understanding the memory layout will help understanding the function.
+
+``NDArray`` has a, and only has a, pointer to a ``Container`` class.
+
+The ``Container`` class is a wrapper for the core data structure ``DLTensor``,
+containing a ``DLTensor`` instance instead of a pointer, together with some
+other variables, providing useful features like reference count and lazy copy.
+
+The memory layout is as below:
+::
+                 
+  +-NDArray----+        +-Container--+
+  | Container*-|------->| +DLTensor+ |      +-----------+
+  +------------+        | | data*--|-|----->|  Memory   |
+                        | | ndim   | |      |           |
+                        | | context| |      |           |
+                        | |  ...   | |      +-----------+
+                        | +--------+ |
+                        |  ref_cnt   |
+                        |  type_code |
+                        |    ...     |
+                        +------------+
+
+This graph shows the memory layout of NDArray.
+
+A ``DLTensor`` is a data structure that holds the data, running context, shape
```

Comment:
The data context instead of `running context`. Cause the `ctx` indicates where the
device memory is allocated.

In fact, you can introduce each element respectively with `list` syntax.